crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,327 @@
+            },
+            ..Default::default()
+        };
+
+        eval_analysis_command
+            .run_json(&nix, &Default::default())
+            .map_err(|nix_error| {
+                PublishError::DrvMetadata(
+                    self.attr_path.clone(),
+                    self.publish_ref.clone(),
+                    nix_error,
+                )
+            })
+            .await
+    }
+
+    /// Resolve the metadata of the flake holding the published package
+    async fn get_flake_metadata(&self) -> PublishResult<FlakeMetadata> {
+        let nix: NixCommandLine = self.flox.nix(Default::default());
+
+        let locked_ref_command = runix::command::FlakeMetadata {
+            flake_ref: Some(self.publish_ref.clone().into_inner().into()),
+            ..Default::default()
+        };
+
+        locked_ref_command
+            .run_typed(&nix, &Default::default())
+            .map_err(|nix_err| PublishError::FlakeMetadata(self.publish_ref.clone(), nix_err))
+            .await
+    }
+}
+
+impl<'flox> Publish<'flox, NixAnalysis> {
+    /// copy the outputs and dependencies of the package to binary store
+    pub async fn upload_binary(self) -> PublishResult<Publish<'flox, NixAnalysis>> {
+        todo!()
+    }
+
+    #[allow(unused)] // until implemented
+    async fn get_binary_cache_metadata(
+        &self,
+        substituter: SubstituterUrl,
+    ) -> PublishResult<CacheMeta> {
+        todo!()
+    }
+
+    /// write snapshot to catalog and push to origin
+    pub async fn push_catalog(self) -> PublishResult<()> {
+        let url = self.publish_ref.clone_url();
+        let repo_dir = tempfile::tempdir_in(&self.flox.temp_dir)
+            .unwrap()
+            .into_path(); // todo catch error
+        let catalog = <Git as GitProvider>::clone(url, &repo_dir, false)
+            .await
+            .unwrap(); // todo: catch error
+
+        if catalog.list_branches().await.unwrap() // todo: catch error
+            .into_iter().any(|info| info.name == "catalog")
+        {
+            catalog.checkout("catalog", false).await.unwrap(); // todo: catch error
+        } else {
+            todo!();
+            // catalog.checkout("catalog", true).await.unwrap(); // todo: catch error
+
+            // catalog.set_upstream("origin", "catalog").await.unwrap();  // todo: implement
+            //                                                               todo: catch error
+        }
+        todo!()
+    }
+
+    /// read out the current publish state
+    pub fn analysis(&self) -> &Value {
+        self.analysis.deref()
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum PublishError {
+    #[error("Failed to load metadata for the package '{0}' in '{1}': {2}")]
+    DrvMetadata(AttrPath, PublishRef, NixCommandLineRunJsonError),
+
+    #[error("Failed to load metadata for flake '{0}': {1}")]
+    FlakeMetadata(PublishRef, NixCommandLineRunJsonError),
+}
+
+type PublishResult<T> = Result<T, PublishError>;
```

Comment:
Honestly it's more common to just write out `Result<T, PublishError>` or say `type Result<T> = Result<T, PublishError>`. The latter is what `std` does for I/O errors.

Reply:
I'm happy to rename this either way, since in this crate were using `Result<T, ___Error>` so far lets forego the treewide renaming and use that style here too